src/test/csfle.rs

```diff
@@ -0,0 +1,1192 @@
+    if !check_env("bson_size_limits", false) {
+        return Ok(());
+    }
+    let _guard = LOCK.run_exclusively().await;
+
+    // Setup: db initialization.
+    let (client, datakeys) = init_client().await?;
+    client
+        .database("db")
+        .create_collection(
+            "coll",
+            CreateCollectionOptions::builder()
+                .validator(doc! { "$jsonSchema": load_testdata("limits-schema.json")? })
+                .build(),
+        )
+        .await?;
+    datakeys
+        .insert_one(load_testdata("limits-key.json")?, None)
+        .await?;
+
+    // Setup: encrypted client.
+    let auto_enc_opts = AutoEncryptionOptions::builder()
+        .key_vault_namespace(KV_NAMESPACE.clone())
+        .kms_providers(LOCAL_KMS.clone())
+        .extra_options(EXTRA_OPTIONS.clone())
+        .build();
+    let client_encrypted =
+        Client::with_encryption_options(CLIENT_OPTIONS.get().await.clone(), auto_enc_opts).await?;
+    let coll = client_encrypted
+        .database("db")
+        .collection::<Document>("coll");
+
+    // Tests
+    let mut value = String::new();
+    // Manually push characters; constructing the array directly explodes the stack.
+    for _ in 0..2097152 {
+        value.push('a');
+    }
+    coll.insert_one(
+        doc! {
+            "_id": "over_2mib_under_16mib",
+            "unencrypted": value,
+        },
+        None,
+    )
+    .await?;
+
+    let mut doc: Document = load_testdata("limits-doc.json")?;
+    doc.insert("_id", "encryption_exceeds_2mib");
+    let mut value = String::new();
+    for _ in 0..(2097152 - 2000) {
+        value.push('a');
+    }
+    doc.insert("unencrypted", value);
+    coll.insert_one(doc, None).await?;
+
+    // TODO(RUST-583) Test bulk write limit behavior
+
+    let mut value = String::new();
+    for _ in 0..(16777216 - 2000) {
+        value.push('a');
+    }
+    let doc = doc! {
+        "_id": "under_16mib",
+        "unencrypted": value,
+    };
+    coll.insert_one(doc, None).await?;
+
+    let mut doc: Document = load_testdata("limits-doc.json")?;
+    doc.insert("_id", "encryption_exceeds_16mib");
+    let mut value = String::new();
+    for _ in 0..(16777216 - 2000) {
+        value.push('a');
+    }
+    doc.insert("unencrypted", value);
+    let result = coll.insert_one(doc, None).await;
+    assert!(result.is_err());
+
+    Ok(())
+}
+
+#[cfg_attr(feature = "tokio-runtime", tokio::test)]
+#[cfg_attr(feature = "async-std-runtime", async_std::test)]
+async fn views_prohibited() -> Result<()> {
+    if !check_env("views_prohibited", false) {
+        return Ok(());
+    }
+    let _guard = LOCK.run_exclusively().await;
+
+    // Setup: db initialization.
+    let (client, _) = init_client().await?;
+    client
+        .database("db")
+        .collection::<Document>("view")
+        .drop(None)
+        .await?;
+    client
+        .database("db")
+        .run_command(doc! { "create": "view", "viewOn": "coll" }, None)
```

Comment:
Yeah, it might be more ergonomic to have separate helpers for creating an actual collection and creating a view. Unfortunately, it looks like there's only tests for collection management but not actually a spec (DRIVERS-2118). Could you file a RUST ticket for this, and maybe when 3.0 comes around it might be enough inertia for one of us to just write the drivers spec for it.

Reply:
Filed RUST-1512 for this.